### DISCUSSION TITLE
Python Crashes with `Windows fatal exception: access violation` when trying to load a pybind11 module

### DISCUSSION THREAD

**Discussion by eschan145** [pybind/pybind11#5442] [Q&A]
Hello all,

I've attempted to use pybind11 on a Windows machine with g++:
```
g++ -ID:/a/_temp/msys64/mingw64/include/python3.12 -c src/util.cpp -o src/util.o -std=c++20
g++ -shared -ID:/a/_temp/msys64/mingw64/include/python3.12 -LD:/a/_temp/msys64/mingw64/lib src/util.o -lpython3.12 -o src/util.pyd
```

However, when I tried to import the generated `pyd` module in my Python code with a simple `import util`, Python crashes silently, without giving an error. I've specified DLLs with `os.add_dll_directory()`. I've also used `faulthandler.enable()` to try to see what is going on, as there is no error output and Python crashes mysteriously. This is what is spat out by fauthandler before the program finishes:

```
Windows fatal exception: access violation

Current thread 0x0000ae5c (most recent call first):
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 1289 in create_module
  File "<frozen importlib._bootstrap>", line 813 in module_from_spec
  File "<frozen importlib._bootstrap>", line 921 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1331 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
```

I doubt the issue is with my code, and rather the environment or my compilation command — running any of the pybind11 examples has the same error.

It's also unlikely that this is due to missing dependencies, as I ran into the issue `The specified module could not be found` and fixed it by using `os.add_dll_directory`.

Thanks for your help in advance.

**Comment by zhanghq-chn** [comment]
I had a very similar problem on my Mac, using CPython C API. The faulthandler spits out basically the same thing.
